src/reconcile/rec_visibility.c

```diff
@@ -270,6 +270,17 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
         if ((txnid = upd->txnid) == WT_TXN_ABORTED)
             continue;
 
+        /*
+         * If we've spotted the update that we last wrote to the data store, then there's no use
+         * continuing. We should just leave that update on the disk.
+         */
+        if (!saw_on_disk_upd && F_ISSET(upd, WT_UPDATE_DS)) {
+            if (F_ISSET_ATOMIC(page, WT_PAGE_LAST_REC_FAILED))
```

Comment:
@bvpvamsikrishna 
Maybe I misunderstood your intention, but I don't this flag is enough.
In the presence of evictions failures, this is still wrong:

- If eviction failed, it may have failed before getting to this key. In that case, the update marked with `WT_UPDATE_DS` may be the real on-disk value and we should have done a `break` here.
- If eviction failed after it got to this key, we are correct to look past it. But there may be another update in the chain marked with `WT_UPDATE_DS` from a previous successful reconciliation that we should NOT look past and we don't enforce this.

I'll have to think about this more but I don't see an easy solution here.

Reply:
My idea is something as below:
1. start of the reconcilation, bool prev_rec_failed = F_ISSET(WT_PAGE_LAST_REC_FAILED)
2. Traverse update chain
    2.1 Examine first visible update
          2.1.1 If update is marked as either WT_UPDATE_DS | WT_UPDATE_HS, 
                  if ( !prev_rec_failed)
                         skip this update chain.
                  else
                        generate page image with upd marked with WT_UPDATE_DS.
          2.1.2 else mark this update as selected update, 
                          and traverse till ( end or update with flag WT_UPDATE_DS).
                             copy the update to HS,  F_CLR (WT_UPDATE_DS), F_SET(WT_UPDATE_HS).
         generate the new page image.
         F_CLR(WT_PAGE_LAST_REC_FAILED)

We can discuss see whether there are any cases which are not handled.

Reply:
de71a51